functorch/_src/aot_autograd.py

```diff
@@ -159,31 +160,83 @@ def forward(ctx, *flat_tensor_args):
+                    to_be_saved = fw_outs[num_outs:] + list(flat_tensor_args) + out
+                    print("fw outs: ", fw_outs, "-------")
+                    ctx.save_for_backward(*to_be_saved)
+                    ctx.fwd_graph = fw_module.code
+                else:
+                    nonlocal compiled_bw
+                    bw_args = fw_outs[num_outs:] + fw_outs[0:num_outs]
+                    compiled_bw = bw_compiler(bw_module, bw_args)
+                    ctx.save_for_backward(*fw_outs[num_outs:])
             else:
                 fw_outs = normalize_as_list(compiled_fw(*flat_tensor_args))
-            ctx.save_for_backward(*fw_outs[num_outs:])
+                if partition_fn is default_partition:
+                    with torch.set_grad_enabled(grad_state):
+                        out = flat_fn(*flat_tensor_args)
+                    out = pytree.tree_map(
+                        lambda x: x.detach().contiguous() if isinstance(x, Tensor) else x, out
+                    )
+                    ctx.num_intermediate = len(fw_outs[num_outs:])
+                    ctx.num_inputs = len(flat_tensor_args)
+                    to_be_saved = fw_outs[num_outs:] + list(flat_tensor_args) + out
+                    ctx.save_for_backward(*to_be_saved)
+                else:
+                    ctx.save_for_backward(*fw_outs[num_outs:])
             return tuple(fw_outs[0:num_outs])
 
         @staticmethod
         @disable_torchdynamo
-        def backward(ctx, *flat_args):
-            contiguous_args = [t.contiguous() for t in flat_args]
-            # contiguous_args = [t for t in flat_args]
-            out = normalize_as_list(compiled_bw(*ctx.saved_tensors, *contiguous_args))
-            return tuple(out)
+        def backward(ctx, *flat_grad_outs):
+            print(flat_grad_outs)
+            contiguous_args = [t.contiguous() for t in flat_grad_outs]
+            if compiled_bw is None:
+                assert partition_fn is default_partition
+                with torch.set_grad_enabled(grad_state):
+                    inputs = ctx.saved_tensors[ctx.num_intermediate:ctx.num_intermediate+ctx.num_inputs]
+                    fx_g = make_fx(joint_forward_backward, aot_decompositions)(inputs, contiguous_args)
+                    # assert that the forward graph generated here is the same
+                    # if it's specified that the user might want to calculate double backwards
+                fw_module, bw_module = partition_fn(fx_g, ctx.saved_tensors[ctx.num_intermediate:])
+                print(fw_module.code)
+                print(ctx.fwd_graph)
+                assert fw_module.code == ctx.fwd_graph
+                func_code = bw_module.code.split('self, ')
+                # print(func_code[0] + func_code[1])
+                exec(func_code[0] + func_code[1], globals())
+                f = create_aot_autograd_function(forward, bw_compiler, bw_compiler, partition_fn, aot_decompositions, grad_state)
```

Comment:
Two questions:
- Why are we passing `forward` to create_aot_autograd_function? I would have expected us to pass bw_module.code without the self argument
- What is the exec for? Are you trying to test this without the `create_aot_autograd_function` line?

Reply:
- forward is the name of the function generated by running `bw_module.code`
- exec executes the bw_module.code to create a backward function which is the forward for the next pass